ui/source/js/components/game.js

```diff
@@ -1,78 +1,127 @@
 	constructor(props) {
 		super(props);
 		this.state = {
-			history: [{
-				pieces: Array(9).fill(null)
-			}],
-			xIsNext: true,
-			stepNumber: 0
+			isBlueTurn: true,
+			score: {
+				black: blackPieceCount,
+				blue: bluePiecesCount,
+				red: redPiecesCount
+			},
+			pieces: buildPieceArray(),
+			viewer: 0,
+			winner: null,
+			teams: generateTeams()
 		};
 	}
 
-	handleClick(i) {
-		const history = this.state.history.slice(0, this.state.stepNumber + 1);
-		const current = history[history.length - 1];
-		const pieces = current.pieces.slice();
-		if (calculateWinner(pieces) || pieces[i]) {
-			return;
-		}
-		pieces[i] = this.state.xIsNext ? 'X' : 'O';
-		this.setState({
-			history: history.concat([{
-				pieces: pieces
-			}]),
-			stepNumber: history.length,
-			xIsNext: !this.state.xIsNext
-		});
+	switchTeamsClick() {
+		this.setState({ isBlueTurn: !this.state.isBlueTurn });
 	}
 
-	jumpTo(step) {
-		this.setState({
-			stepNumber: step,
-			xIsNext: (step % 2) === 0,
-		});
+	setViewer(i) {
+		this.setState({ viewer: i });
 	}
 
-	render() {
-		const history = this.state.history;
-		const current = history[this.state.stepNumber];
-		const winner = calculateWinner(current.pieces);
+	pieceChosenClick(i) {
+		if (this.state.winner === null) {
+			const updatedPieces = Array.from(this.state.pieces);
+			const clickedPieceOwner = updatedPieces[i].owner;
+			updatedPieces[i].isChosen = true;
+			const newScore = keepScore(updatedPieces, blackPieceCount, bluePiecesCount, redPiecesCount);
+			const newWinner = calculateWinner(newScore, this.state.isBlueTurn);
+			this.setState({
+				score: newScore,
+				pieces: updatedPieces,
+				winner: newWinner,
+				isBlueTurn: switchTurns(this.state.isBlueTurn, clickedPieceOwner)
+			});
+			if (newWinner !== null) {
+				this.endGame(newWinner);
+			}
+		}
+		return false;
+	}
+
+	promptNewGame() {
+		if (confirm(strResetGame)) {
+			this.newGame();
+		}
+	}
 
-		const moves = history.map((step, move) => {
-			const desc = move ?
-				'Go to move #' + move :
-				'Go to game start';
-			return (
-				<li key={move}>
-					<button onClick={() => this.jumpTo(move)}>{desc}</button>
-				</li>
-			);
+	newGame() {
+		this.setState({
+			winner: null,
+			isBlueTurn: true,
+			pieces: buildPieceArray()
 		});
+	}
 
-		let status;
-		if (winner) {
-			status = 'Winner: ' + winner;
-		} else {
-			status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
-		}
+	endGame() {
+		const updatedPieces = Array.from(this.state.pieces);
+		updatedPieces.forEach(piece => piece.isChosen = true);
+		this.setState({ pieces: updatedPieces });
+	}
 
+	render() {
 		return (
 			<div className="game">
-				<div className="game-board">
-					<Board
-						pieces={current.pieces}
-						onClick={(i) => this.handleClick(i)}
-					/>
-				</div>
-				<div className="game-info">
-					<div>{status}</div>
-					<ol>{moves}</ol>
-				</div>
+				<Stats
+					isBlueTurn={this.state.isBlueTurn}
```

Comment:
Something to keep in mind as you continue with React: destructuring will save you a lot of keystrokes. Doing something like:
`const { isBlueTurn, winner, viewer } = this.state;`
will save you from having to type `this.state.winner`, `this.state.viewer`, etc a bunch of times.